Change ESM build to generate modules for each input file for better treeshaking

Querela opened the pull request:
Implementing https://github.com/mike-lischke/antlr4ng/issues/103

This pull requests improves tree-shaking for using `antlr4ng` with ESM modules.

It add a `esbuild` script `build/build-esm.mjs` that will traverse the import tree from a given input script and runs the ESM build with `esbuild` for each file separately. This will generate for each source input file an output file.

The `package.json` configuration includes the `"sideEffects": false` statement.
The original single bundle `index.mjs` will now contain imports to each module that `antlr4ng` requires not not be a single bundle. All the ESM/ES6 module files have the `.mjs` extension.

mike-lischke commented:
Can't we get the same result by just running:

```
        "build-mjs2": "esbuild ./src/**/*.ts ./src/index.ts --bundle --main-fields=module,main --target=esnext --keep-names --outdir=dist/ --format=esm --out-extension:.js='.mjs'",
        "build-cjs2": "esbuild ./src/**/*.ts ./src/index.ts --bundle --main-fields=module,main --target=esnext --keep-names --outdir=dist/ --format=cjs --out-extension:.js='.cjs'",
```

? At least that generates the individual cjs and mjs files. Unfortunately, this fails to import in a project which uses these files, but it could be that's just due to the way I tested that (using npm link).

mike-lischke commented:
In another project it worked nicely (where I don't mix antlr4ng packages (one via npm link and one imported as usual). So, I'm not sure we need this PR. Can you pls test that?

Querela commented:
Yes, I will test this (tomorrow probably). 😃 
I tried something similar with a variant of the command already but were some issues (e.g., imports referrring to `.js` but files using `.mjs`; glob pattern requiring quoting; I also did not use the `--bundle` param). So, I will be happy if this one works.

Querela commented:
So, I tested it out with a fresh clone of `antlr4ng`. 
Two observations. The commands work but not as expected:
- The parameter `--bundle` will still create bundles that include imports transitively (I think), so if you inspect the results in `dist/`, you will still find that each file can contain multiple sources (recognizable from the `// src/...` comments). So, direct imports of modules in user code are self-contained (they contain everything they import) but is a lot of code duplication, especially if users will import multiple modules. Might then have other issues with identity.
- The glob (`./src/**/*.ts`) needs to be quoted. _I did this mistake, too._ Without quoting it uses the shell's globbing, not esbuild's. I'm a bit confused why it will still create the `dist/index.mjs` besides all the other 1-directory files. But it notably did not create output files for directory of the root level and level 2+, see the missing `dist/tree/xpath` and `dist/tree/pattern` folders/files.

The updated command `esbuild './src/**/*.ts' ./src/index.ts --main-fields=module,main --target=esnext --keep-names --outdir=dist/ --format=esm --out-extension:.js='.mjs'` would be better but suffers the issues with the `out-extensions`. Files are correctly renamed but imports are not updated. So, a `const module = await import("./dist/index.mjs");` will throw in `node`.
I also tried it with `--bundle --external:*` but this also does not correctly update the import filename extensions. My idea was that it will try to run in a bundling mode but externalize all imports so they are not bundled.

And ... I'm unsure whether CommonJS requires this file splitting. It will probably benefit from reduced transitive imports, I think, but will as a consequence have lots of synchronous import processes running which might make it slower. A single bundle might be more optimal here? "Literature" (blogs, documentation) so far only mention tree-shaking for ESM and the multi-file builds.

_Another major annoyance it that a lot of tree-shaking articles primarily refer to scenarios where the author wants to create a minimal bundle. It mentions library builds but I feel the context is often incorrect. E.g., you - as the `antlr4ng` author - already have a clean library where you don't need and want to tree-shake unnecessary code (code editors already highlight unused/dead code anyway). So, the focus of articles is less on (a) tree-shaking third-party dependencies as it either works or doesn't automatically, and (b) creating a library that can be tree-shaken (and is not already tree-shaking in the build process)...
/rant over_

mike-lischke commented:
> The parameter --bundle will still create bundles that include imports transitively (I think), so if you inspect the results in dist/, you will still find that each file can contain multiple sources (recognizable from the // src/... comments). 

Oh, I didn't see that, but if you think about it, it makes sense. The build uses a file as start and creates the bundle. So, if you give it multiple files, it will create multiple bundles. ¯\_(ツ)_/¯

But that invalids my test actually. It probably worked only because whatever file I imported contained the entire bundle.

> The glob (./src/**/*.ts) needs to be quoted. I did this mistake, too. Without quoting it uses the shell's globbing, not esbuild's. 

Dang, all these little pitfalls. But does this change the behavior?

> The updated command esbuild './src/**/*.ts' ./src/index.ts --main-fields=module,main --target=esnext --keep-names --outdir=dist/ --format=esm --out-extension:.js='.mjs' would be better but suffers the issues with the out-extensions. Files are correctly renamed but imports are not updated. So, a const module = await import("./dist/index.mjs"); will throw in node.

Well, that's all the result of not bundling the files! I told you it has side effects. However, it would be enough to name the top level index file properly, I think. it should be transformed to import dependent files either using ESM or CJS.

> And ... I'm unsure whether CommonJS requires this file splitting. It will probably benefit from reduced transitive imports, I think, but will as a consequence have lots of synchronous import processes running which might make it slower. A single bundle might be more optimal here? "Literature" (blogs, documentation) so far only mention tree-shaking for ESM and the multi-file builds.

... which makes me think we shouldn't further investigate here.

On the other hand, maybe the problem lies in a totally different area: I use index files import and re-export class, to allow using a single `from "antlr4ng"` import, without paths. This is a reminiscense of the time before I used esbuild. Maybe removing these index files solves the tree-shaking problem?

mike-lischke commented:
This supports my idea: https://github.com/evanw/esbuild/issues/1420

Querela commented:
So, I did not have as much time as I planned to investigate further, might be next week at the earliest...

> > The parameter --bundle will still create bundles that include imports transitively (I think), so if you inspect the results in dist/, you will still find that each file can contain multiple sources (recognizable from the // src/... comments).
> 
> Oh, I didn't see that, but if you think about it, it makes sense. The build uses a file as start and creates the bundle. So, if you give it multiple files, it will create multiple bundles. `¯\_(ツ)_/¯`
> 
> But that invalids my test actually. It probably worked only because whatever file I imported it contained the entire bundle.

I think so, yes.

> > The glob (./src/**/*.ts) needs to be quoted. I did this mistake, too. Without quoting it uses the shell's globbing, not esbuild's.
> 
> Dang, all these little pitfalls. But does this change the behavior?

It breaks stuff due to the extensions. It will generate files 1:1 so `*.ts` -> `*.mjs` but imports will fail.

> > The updated command esbuild './src/**/*.ts' ./src/index.ts --main-fields=module,main --target=esnext --keep-names --outdir=dist/ --format=esm --out-extension:.js='.mjs' would be better but suffers the issues with the out-extensions. Files are correctly renamed but imports are not updated. So, a const module = await import("./dist/index.mjs"); will throw in node.
> 
> Well, that's all the result of not bundling the files! I told you it has side effects. However, it would be enough to name the top level index file properly, I think. it should be transformed to import dependent files either using ESM or CJS.

I believe I have this fixed with the updated pull request, where I refactored and extracted the build logic into a plugin that (a) creates output files for each imported module and (b) updates the imports to make it work.

> > And ... I'm unsure whether CommonJS requires this file splitting. It will probably benefit from reduced transitive imports, I think, but will as a consequence have lots of synchronous import processes running which might make it slower. A single bundle might be more optimal here? "Literature" (blogs, documentation) so far only mention tree-shaking for ESM and the multi-file builds.
> 
> ... which makes me think we shouldn't further investigate here.

It's not high priority, yes.

> On the other hand, maybe the problem lies in a totally different area: I use index files to import and re-export classes, to allow using a single `from "antlr4ng"` import, without paths. This is a reminiscense of the time before I used esbuild. Maybe removing these index files solves the tree-shaking problem?

I don't think this is the problem, and it would make using the library a bit harder. Especially the generated code seems to rely on the `antlr.` namespace (`import * as antlr from "antlr4ng";`), so changing would break too much. And other libraries I took a look at still use re-exports without affecting tree-shaking.

Querela commented:
Ok. I still do not have time for the above at the moment, but I noticed an issue due to https://github.com/mike-lischke/antlr4ng/commit/00e74ee6286041ca954d6a91ffd2216bb2dd0982. I was wondering why it now used the `index.cjs` instead of `index.mjs` in my test setup.
It seems the order of "exports" matters, so the catch-all `default` before the `module` field will block it from being used. So, `default` should be the last element.
Who would have known ... It's not like dictionaries have a guaranteed order like arrays?! _Not sure if this is a vite/rollup/esbuild/... issue when building my application._

I also do not see it [mentioned here](https://hirok.io/posts/package-json-exports#targeting-node-js-browser) for example but it seems to matter... Curious that the [vite docs](https://vite.dev/guide/build#library-mode) also still use the `require`/`import` forms in their examples.

mike-lischke commented:
I just tried that again and found it not working at all, so I'm going back to the original exports object content.
